### PR TITLE
Added Vector store for PGVector

### PR DESCRIPTION
**Pull request by AyoTheDev**

### CODE REVIEW

**Comment by li2109** [comment]
> Kudos, SonarCloud Quality Gate passed!    [Quality Gate passed](https://sonarcloud.io/dashboard?id=Knowly-ai_langtorch&pullRequest=134)
> 
> [Bug](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=BUG) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=BUG) [0 Bugs](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=BUG) [Vulnerability](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=VULNERABILITY) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=VULNERABILITY) [0 Vulnerabilities](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=VULNERABILITY) [Security Hotspot](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=SECURITY_HOTSPOT) [A](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=SECURITY_HOTSPOT) [0 Security Hotspots](https://sonarcloud.io/project/security_hotspots?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=SECURITY_HOTSPOT) [Code Smell](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=CODE_SMELL) [A](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=CODE_SMELL) [19 Code Smells](https://sonarcloud.io/project/issues?id=Knowly-ai_langtorch&pullRequest=134&resolved=false&types=CODE_SMELL)
> 
> [](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=134&metric=new_coverage&view=list) [0.0% Coverage](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=134&metric=new_coverage&view=list) [](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=134&metric=new_duplicated_lines_density&view=list) [0.0% Duplication](https://sonarcloud.io/component_measures?id=Knowly-ai_langtorch&pullRequest=134&metric=new_duplicated_lines_density&view=list)

Please resolve the code smell

**Comment by li2109** [comment]
A general approach to error handling:
use checked exceptions for conditions from which the caller can reasonably be expected to recover.
Use runtime exceptions to indicate programming errors.

**Comment by li2109** [comment]
Let's add comments for every public class

**Review by li2109 — CHANGES_REQUESTED** [review]

File: src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/schema/PGVectorStoreSpec.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import java.sql.Connection;
+import java.util.Optional;
+
+@Data
+@Builder(toBuilder = true, setterPrefix = "set")
+public class PGVectorStoreSpec {
+
+    @NonNull private final PGVectorService pgVectorService;
+    @Builder.Default private final String model = "text-embedding-ada-002";
```
[the remark this answers is not in the document]
Comment (AyoTheDev) [reply src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/schema/PGVectorStoreSpec.java:18]: Do you mean to create an EmbeddingModelConstants class that will contain String Constants like "text-embedding-ada-002"?

File: src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/PGVectorStore.java [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+        return Pair.of(document, documentWithScore.getRight());
+    }
+
+    private float[] getFloatVectorValues(List<Double> vectorValues) {
+        float[] floats = new float[vectorValues.size()];
+        int i = 0;
+        for (Double vectorValue : vectorValues) {
+            floats[i] = vectorValue.floatValue();
+            i++;
+        }
```
[the remark this answers is not in the document]
Comment (AyoTheDev) [reply src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/PGVectorStore.java:328]: We can't directly convert a float values to double

File: src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/PGVectorStore.java [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        return floats;
+    }
+
+    private double[] getDoubleVectorValues(float[] vectorValues) {
+        double[] doubles = new double[vectorValues.length];
+        for (int i = 0; i < vectorValues.length; i++) {
+            doubles[i] = vectorValues[i];
+        }
+        return doubles;
```
[the remark this answers is not in the document]
Comment (AyoTheDev) [reply src/main/java/ai/knowly/langtorch/store/vectordb/integration/pgvector/PGVectorStore.java]: We can't directly convert float values to double

**Comment by li2109** [comment]
Wow, the PR is well-documented!!!

**Comment by li2109** [comment]
IDK why some comments are left here https://github.com/AyoTheDev/langtorch/pull/1

**Pull request merged**